File: ring-android/libringclient/src/main/java/net/jami/account/JamiAccountSummaryPresenter.java
/*
 *  Copyright (C) 2004-2021 Savoir-faire Linux Inc.
 *
 *  Author: Alexandre Lision <alexandre.lision@savoirfairelinux.com>
 *  Author: Adrien Béraud <adrien.beraud@savoirfairelinux.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package net.jami.account;

import net.jami.model.Account;
import net.jami.services.AccountService;
import net.jami.services.DeviceRuntimeService;

import java.io.File;
import java.net.SocketException;

import javax.inject.Inject;
import javax.inject.Named;

import net.jami.mvp.RootPresenter;
import net.jami.utils.Log;
import net.jami.utils.StringUtils;
import net.jami.utils.VCardUtils;

import ezvcard.property.Photo;
import ezvcard.property.RawProperty;
import ezvcard.property.Uid;
import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class JamiAccountSummaryPresenter extends RootPresenter<JamiAccountSummaryView> {

    private static final String TAG = JamiAccountSummaryPresenter.class.getSimpleName();

    private final net.jami.services.DeviceRuntimeService mDeviceRuntimeService;
    private final net.jami.services.AccountService mAccountService;
    private String mAccountID;

    @Inject
    @Named("UiScheduler")
    protected Scheduler mUiScheduler;

    @Inject
    public JamiAccountSummaryPresenter(AccountService accountService,
                                       DeviceRuntimeService deviceRuntimeService) {
        mAccountService = accountService;
        mDeviceRuntimeService = deviceRuntimeService;
    }

    public void registerName(String name, String password) {
        final Account account = mAccountService.getAccount(mAccountID);
        if (account == null || getView() == null) {
            return;
        }
        mAccountService.registerName(account, password, name);
        getView().accountChanged(account);
    }

    public void startAccountExport(String password) {
        if (getView() == null) {
            return;
        }
        getView().showExportingProgressDialog();
        mCompositeDisposable.add(mAccountService
                .exportOnRing(mAccountID, password)
                .observeOn(mUiScheduler)
                .subscribe(pin -> getView().showPIN(pin),
                           error -> {
                    if (error instanceof IllegalArgumentException) {
                        getView().showPasswordError();
                    } else if (error instanceof SocketException) {
                        getView().showNetworkError();
                    } else {
                        getView().showGenericError();
                    }
                }));
    }

    public void setAccountId(String accountID) {
        mCompositeDisposable.clear();
        mAccountID = accountID;
        net.jami.account.JamiAccountSummaryView v = getView();
        Account account = mAccountService.getAccount(mAccountID);
        if (v != null && account != null)
            v.accountChanged(account);
        mCompositeDisposable.add(mAccountService.getObservableAccountUpdates(mAccountID)
                .observeOn(mUiScheduler)
                .subscribe(a -> {
                    net.jami.account.JamiAccountSummaryView view = getView();
                    if (view != null)
                        view.accountChanged(a);
                }));
    }

    public void enableAccount(boolean newValue) {
        Account account = mAccountService.getAccount(mAccountID);
        if (account == null) {
            Log.w(TAG, "account not found!");
            return;
        }

        account.setEnabled(newValue);
        mAccountService.setAccountEnabled(account.getAccountID(), newValue);
    }

    public void changePassword(String oldPassword, String newPassword) {
        net.jami.account.JamiAccountSummaryView view = getView();
        if (view != null)
            view.showPasswordProgressDialog();
        mCompositeDisposable.add(mAccountService.setAccountPassword(mAccountID, oldPassword, newPassword)
                .observeOn(mUiScheduler)
                .subscribe(
                        () -> getView().passwordChangeEnded(true),
                        e -> getView().passwordChangeEnded(false)));
    }

    public String getDeviceName() {
        Account account = mAccountService.getAccount(mAccountID);
        if (account == null) {
            Log.w(TAG, "account not found!");
            return null;
        }
        return account.getDeviceName();
    }

    public void downloadAccountsArchive(File dest, String password) {
        getView().showExportingProgressDialog();
        mCompositeDisposable.add(
            mAccountService.exportToFile(mAccountID, dest.getAbsolutePath(), password)
            .observeOn(mUiScheduler)
            .subscribe(() -> getView().displayCompleteArchive(dest),
                    error -> getView().passwordChangeEnded(false)));
    }

    public void saveVCardFormattedName(String username) {
        Account account = mAccountService.getAccount(mAccountID);
        File filesDir = mDeviceRuntimeService.provideFilesDir();

        mCompositeDisposable.add(VCardUtils.loadLocalProfileFromDiskWithDefault(filesDir, mAccountID)
                .doOnSuccess(vcard -> {
                    vcard.setFormattedName(username);
                    vcard.removeProperties(RawProperty.class);
                })
                .flatMap(vcard -> VCardUtils.saveLocalProfileToDisk(vcard, mAccountID, filesDir))
                .subscribeOn(Schedulers.io())
                .subscribe(vcard -> {
                    account.resetProfile();
                    mAccountService.refreshAccounts();
                    getView().updateUserView(account);
                }, e -> Log.e(TAG, "Error saving vCard !", e)));
    }

    public void saveVCard(String username, Single<Photo> photo) {
        Account account = mAccountService.getAccount(mAccountID);
        String ringId = account.getUsername();
        File filesDir = mDeviceRuntimeService.provideFilesDir();
        mCompositeDisposable.add(Single.zip(
                VCardUtils.loadLocalProfileFromDiskWithDefault(filesDir, mAccountID).subscribeOn(Schedulers.io()),
                photo, (vcard, pic) -> {
                    vcard.setUid(new Uid(ringId));
                    if (!StringUtils.isEmpty(username)) {
                        vcard.setFormattedName(username);
                    }
                    vcard.removeProperties(Photo.class);
                    vcard.addPhoto(pic);
                    vcard.removeProperties(RawProperty.class);
                    return vcard;
                })
                .flatMap(vcard -> VCardUtils.saveLocalProfileToDisk(vcard, mAccountID, filesDir))
                .subscribeOn(Schedulers.io())
                .subscribe(vcard -> {
                    account.resetProfile();
                    mAccountService.refreshAccounts();
                    getView().updateUserView(account);
                }, e -> Log.e(TAG, "Error saving vCard !", e)));
    }

    public void cameraClicked() {
        boolean hasPermission = mDeviceRuntimeService.hasVideoPermission() &&
                mDeviceRuntimeService.hasWriteExternalStoragePermission();
        net.jami.account.JamiAccountSummaryView view = getView();
        if (view != null) {
            if (hasPermission) {
                view.gotToImageCapture();
            } else {
                view.askCameraPermission();
            }
        }
    }

    public void galleryClicked() {
        boolean hasPermission = mDeviceRuntimeService.hasGalleryPermission();
        if (hasPermission) {
            getView().goToGallery();
        } else {
            getView().askGalleryPermission();
        }
    }

    public void goToAccount() {
        getView().goToAccount(mAccountID);
    }

    public void goToMedia() {
        getView().goToMedia(mAccountID);
    }

    public void goToSystem() {
        getView().goToSystem(mAccountID);
    }

    public void goToAdvanced() {
        getView().goToAdvanced(mAccountID);
    }

    public void revokeDevice(final String deviceId, String password) {
        if (getView() != null) {
            getView().showRevokingProgressDialog();
        }
        mCompositeDisposable.add(mAccountService
                .revokeDevice(mAccountID, password, deviceId)
                .observeOn(mUiScheduler)
                .subscribe(result -> {
                    getView().deviceRevocationEnded(deviceId, result);
                    getView().updateDeviceList(mAccountService.getAccount(mAccountID).getDevices(),
                            mAccountService.getAccount(mAccountID).getDeviceId());
                }));
    }

    public void renameDevice(String newName) {
        mAccountService.renameDevice(mAccountID, newName);
    }

    public Account getAccount() {
        return mAccountService.getAccount(mAccountID);
    }
}

File: ring-android/libringclient/src/main/java/net/jami/account/JamiAccountSummaryPresenter.kt
/*
 *  Copyright (C) 2004-2021 Savoir-faire Linux Inc.
 *
 *  Author: Alexandre Lision <alexandre.lision@savoirfairelinux.com>
 *  Author: Adrien Béraud <adrien.beraud@savoirfairelinux.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package net.jami.account

import ezvcard.VCard
import ezvcard.property.Photo
import ezvcard.property.RawProperty
import ezvcard.property.Uid
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.schedulers.Schedulers
import net.jami.model.Account
import net.jami.mvp.RootPresenter
import net.jami.services.AccountService
import net.jami.services.DeviceRuntimeService
import net.jami.services.VCardService
import net.jami.utils.Log
import net.jami.utils.StringUtils.isEmpty
import net.jami.utils.VCardUtils.loadLocalProfileFromDiskWithDefault
import net.jami.utils.VCardUtils.saveLocalProfileToDisk
import java.io.File
import java.net.SocketException
import javax.inject.Inject
import javax.inject.Named

class JamiAccountSummaryPresenter @Inject constructor(
    private val mAccountService: AccountService,
    private val mDeviceRuntimeService: DeviceRuntimeService,
    private val mVcardService: VCardService,
    @param:Named("UiScheduler") private val mUiScheduler: Scheduler
) : RootPresenter<JamiAccountSummaryView>() {
    private var mAccountID: String? = null

    fun registerName(name: String?, password: String?) {
        val account = mAccountService.getAccount(mAccountID) ?: return
        mAccountService.registerName(account, password, name)
        view?.accountChanged(account)
    }

    fun startAccountExport(password: String?) {
        if (view == null || mAccountID == null) {
            return
        }
        view?.showExportingProgressDialog()
        mCompositeDisposable.add(mAccountService
            .exportOnRing(mAccountID!!, password!!)
            .observeOn(mUiScheduler)
            .subscribe({ pin: String -> view?.showPIN(pin) }) { error: Throwable ->
                when (error) {
                    is IllegalArgumentException -> view?.showPasswordError()
                    is SocketException -> view?.showNetworkError()
                    else -> view?.showGenericError()
                }
            })
    }

    fun setAccountId(accountId: String) {
        mCompositeDisposable.clear()
        mAccountID = accountId
        mAccountService.getAccount(accountId)?.let { account -> view?.accountChanged(account) }
        mCompositeDisposable.add(mAccountService.getObservableAccountUpdates(accountId)
            .observeOn(mUiScheduler)
            .subscribe { a: Account -> view?.accountChanged(a) })
    }

    fun enableAccount(newValue: Boolean) {
        val account = mAccountService.getAccount(mAccountID)
        if (account == null) {
            Log.w(TAG, "account not found!")
            return
        }
        account.isEnabled = newValue
        mAccountService.setAccountEnabled(account.accountID, newValue)
    }

    fun changePassword(oldPassword: String, newPassword: String) {
        view?.showPasswordProgressDialog()
        mCompositeDisposable.add(mAccountService.setAccountPassword(mAccountID!!, oldPassword, newPassword)
            .observeOn(mUiScheduler)
            .subscribe({ view?.passwordChangeEnded(true) })
            { view?.passwordChangeEnded(false) })
    }

    val deviceName: String?
        get() {
            val account = mAccountService.getAccount(mAccountID)
            if (account == null) {
                Log.w(TAG, "account not found!")
                return null
            }
            return account.deviceName
        }

    fun downloadAccountsArchive(dest: File, password: String?) {
        view?.showExportingProgressDialog()
        mCompositeDisposable.add(
            mAccountService.exportToFile(mAccountID!!, dest.absolutePath, password!!)
                .observeOn(mUiScheduler)
                .subscribe({ view?.displayCompleteArchive(dest) })
                { view?.passwordChangeEnded(false) })
    }

    fun saveVCardFormattedName(username: String?) {
        val account = mAccountService.getAccount(mAccountID)
        val filesDir = mDeviceRuntimeService.provideFilesDir()
        mCompositeDisposable.add(loadLocalProfileFromDiskWithDefault(filesDir, mAccountID!!)
            .doOnSuccess { vcard: VCard ->
                vcard.setFormattedName(username)
                vcard.removeProperties(RawProperty::class.java)
            }
            .flatMap { vcard: VCard -> saveLocalProfileToDisk(vcard, mAccountID!!, filesDir) }
            .subscribeOn(Schedulers.io())
            .subscribe({ vcard: VCard -> account?.loadedProfile = mVcardService.loadVCardProfile(vcard).cache() })
            { e: Throwable -> Log.e(TAG, "Error saving vCard !", e) })
    }

    fun saveVCard(username: String?, photo: Single<Photo>) {
        val account = mAccountService.getAccount(mAccountID)!!
        val ringId = account.username
        val filesDir = mDeviceRuntimeService.provideFilesDir()
        mCompositeDisposable.add(Single.zip(
            loadLocalProfileFromDiskWithDefault(filesDir, mAccountID!!).subscribeOn(Schedulers.io()),
            photo, { vcard: VCard, pic: Photo ->
                vcard.uid = Uid(ringId)
                if (!isEmpty(username)) {
                    vcard.setFormattedName(username)
                }
                vcard.removeProperties(Photo::class.java)
                vcard.addPhoto(pic)
                vcard.removeProperties(RawProperty::class.java)
                vcard
            })
            .flatMap { vcard: VCard -> saveLocalProfileToDisk(vcard, mAccountID!!, filesDir) }
            .subscribeOn(Schedulers.io())
            .subscribe({ vcard: VCard -> account.loadedProfile = mVcardService.loadVCardProfile(vcard).cache() })
            { e: Throwable -> Log.e(TAG, "Error saving vCard !", e) })
    }

    fun cameraClicked() {
        val hasPermission = mDeviceRuntimeService.hasVideoPermission() && mDeviceRuntimeService.hasWriteExternalStoragePermission()
        val view = view
        if (view != null) {
            if (hasPermission) {
                view.gotToImageCapture()
            } else {
                view.askCameraPermission()
            }
        }
    }

    fun galleryClicked() {
        val hasPermission = mDeviceRuntimeService.hasGalleryPermission()
        if (hasPermission) {
            view!!.goToGallery()
        } else {
            view!!.askGalleryPermission()
        }
    }

    fun goToAccount() {
        view?.goToAccount(mAccountID!!)
    }

    fun goToMedia() {
        view?.goToMedia(mAccountID!!)
    }

    fun goToSystem() {
        view?.goToSystem(mAccountID!!)
    }

    fun goToAdvanced() {
        view?.goToAdvanced(mAccountID!!)
    }

    fun revokeDevice(deviceId: String?, password: String?) {
        view?.showRevokingProgressDialog()
        mCompositeDisposable.add(mAccountService
            .revokeDevice(mAccountID!!, password!!, deviceId!!)
            .observeOn(mUiScheduler)
            .subscribe { result: Int ->
                val account = mAccountService.getAccount(mAccountID)!!
                view?.deviceRevocationEnded(deviceId, result)
                view?.updateDeviceList(account.devices, account.deviceId)
            })
    }

    fun renameDevice(newName: String) {
        mAccountService.renameDevice(mAccountID!!, newName)
    }

    val account: Account?
        get() = mAccountService.getAccount(mAccountID)

    companion object {
        private val TAG = JamiAccountSummaryPresenter::class.simpleName!!
    }
}